python/BioSimSpace/Sandpit/Exscientia/_SireWrappers/_system.py

```diff
@@ -812,6 +812,34 @@ def nPerturbableMolecules(self):
         """
         return len(self.getPerturbableMolecules())
 
+    def getDecoupledMolecules(self):
+        """Return a list containing all of the decoupled molecules in the system.
+
+           Returns
+           -------
+
+           molecules : [:class:`Molecule <BioSimSpace._SireWrappers.Molecule>`]
+               A list of decoupled molecules.
+        """
+
+        molecules = []
+
+        for mol in self:
+            if mol.isDecoupled():
+                molecules.append(_Molecule(mol))
+        return molecules
```

Comment:
WIth Sire 2023 you _should_ be able to avoid the loop by searching for the property directly, e.g.:

```python
molecules = []
for mol in self.search("property decouple").molecules():
    molecules.append(mol)
```

This will be _much_ faster for large systems.

Reply:
Sorry I'm not very familiar with this API.
So for a system that has decoupled molecule, if I do `self.search("property decouple")`, I got 
```
Traceback (most recent call last):
  File "/Users/zwu/miniconda3_x86/envs/exs-orionbssfloes/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3433, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-10-b8fc0224423d>", line 1, in <module>
    self.search("property decouple")
  File "/Users/zwu/bitbucket/BioSimSpace/python/BioSimSpace/Sandpit/Exscientia/_SireWrappers/_system.py", line 983, in search
    raise ValueError(msg) from None
ValueError: 'Invalid search query: 'property decouple' : 'Nothing matched the search.'
```

The decoupled molecule is there as if I do `self._sire_object[0].has_property('decouple')`, I got True.

```
>>> self._sire_object[0].property('decouple')
Properties(
    LJ => [ 1,0 ],
    intramol => True,
    charge => [ 1,0 ]
)
```

Reply:
Hmm, that's strange. For example, if I do the following:

```python
import BioSimSpace as BSS

s = BSS.IO.readMolecules("amber/ala/*")

# Extract the first molecule and give it some dummy property.
m = s[0]
m._sire_object = m._sire_object.edit().setProperty("test", 1.0).molecule().commit()

# Update the molecule in the sytem.
s.update(m)

# Now search for molecules with the matching property.
s.search("property test").molecules()[0]
<BioSimSpace.Molecule: nAtoms=22, nResidues=3>
```

Reply:
So in this case, `decouple` is not a `int` but a dictionary with a couple keys.
```
Properties(
    LJ => [ 1,0 ],
    intramol => True,
    charge => [ 1,0 ]
)
```
Would this work as well?

Reply:
I've just tried and it does indeed fail when using a dictionary. I'll raise this as a Sire issue since there's no reason why it shouldn't work the same. Leave things as they are for now and we can update when it's working.

Reply:
Lester and I have talked about this - the current property search only supports simple properties (convertible to number or string). This is an excellent use case for supporting more complex properties. I am looking at adding this now to the `feat_2023_0` branch in Sire (as well as me making the cursor work with setting list and dictionary properties). Lester will update this once this is finished and will update once `feat_2023_0` is merged in and it is working.

Reply:
That was quicker and easier than I thought. I've added the search for molecules that have a matching property in [this commit](https://github.com/michellab/Sire/commit/941603ba2dd5ef8fa39e8859e364c6fc8858bd8f) on `feat_2023_0`. It is quite self-contained (just some edits to `idengine.cpp` plus adding in a new file for search unit tests). You could merge this single commit into `devel` if you want access to this earlier.

I also fixed the error when assigning list/dict properties in the [prior commit](https://github.com/michellab/Sire/commit/244bca454f8d76ac0f946e8f6df4ee2e67eb2e6f). This now works as you would expect, e.g. `cursor["list_prop"] = [1, 2, "cat"]`.